File: python/box.py
import socket;
import sys;
import re;

def connect(ip, port):
    status= True
    print "Connecting..."

    try:
        s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    except:
        status = False
        print "Couldn't open socket"

    try:
        s.connect((ip, int(port)));
    except socket.error, msg:
        status = False
        print "Invalid host or port"

    if not status:
        return

    print "Connected to", ip, "at port", port
    read_input(s)
    return True


def read_input(s):
    while True:
        cmd = raw_input('box> ')
        if cmd == "quit":
            s.close()
            break;
        else:
            s.send(cmd + '\r\n')


def get_args():
    argv = sys.argv
    if len(argv) <= 2:
        print "usage: ", argv[0], "<IP> <PORT>"
        return False
    else:
        if validate_args(argv[1], "ip") and validate_args(argv[2], "port"):
            return [argv[1], argv[2]]
        else:
            return False


def validate_args(value, kind):
    ipValid     = False
    portValid   = False

    if kind == "ip":
        ipValid = not not re.match('^\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}$', value)
        return ipValid
    if kind == "port":
        portValid = not not re.match('^\d{1,5}$', value)
        return portValid
    else:
        return False

    return True
        

def init():
    args = get_args()
    if args:
        connect(args[0], args[1])

init()

File: c/box.c
#include <stdio.h>
#include <stdlib.h>
#include "lib/include/utils.h"

//socket
#include <sys/types.h>
#include <sys/socket.h>
#include <netdb.h>
#include <string.h>

#define MAX_CMD_LENGTH  20
#define QUIT            "quit"
#define BUFSIZE         1024

bool box_process_input(char cmd[MAX_CMD_LENGTH]) {
    if(str_cmp(cmd, QUIT)) {
        return FALSE;
    }
    return TRUE;
}

void box_read_input(int sockfd) {
    char cmd[MAX_CMD_LENGTH];

    do {
        fprintf(stdout, "box> ");
        fgets(cmd, MAX_CMD_LENGTH, stdin); // TODO: replace to fscanf
        send(sockfd, cmd, strlen(cmd),0);
    } while(box_process_input(cmd));
    shutdown(sockfd, 2);
}

int box_connect(char *host, int port) {
    struct hostent     *he;
    struct sockaddr_in  server;
    int                 sockfd;

    if((he = gethostbyname(host)) == NULL) {
        perror("error resolving hostname");
        exit(1);
    }

    sockfd = socket(AF_INET, SOCK_STREAM, 0);

    memcpy(&server.sin_addr, he->h_addr_list[0], he->h_length);
    server.sin_family   = AF_INET;
    server.sin_port     = htons(port);

    if(connect(sockfd, (struct sockaddr *) &server, sizeof(server)) == -1) {
        perror("error connecting");
        exit(1);
    }
    return sockfd;
}

int main(int argc, char **argv) {
    int sockfd = box_connect("192.168.1.64", 8082);
    box_read_input(sockfd);
    return 0;
}
